src/sqlfluff/rules/L042.py

```diff
@@ -48,44 +70,281 @@ class Rule_L042(BaseRule):
+                "select_statement",
+            )
+        )
+        for cte in root_select.children(is_type("common_table_expression")):
+            assert isinstance(cte, CTEDefinitionSegment), "TypeGaurd"
+            ctes.insert_cte(cte)
+
+    lint_results: List[LintResult] = []
+    CTESegment = Seg(CTEDefinitionSegment)
+    TableExpressionSeg = Seg(TableExpressionSegment)
+    TableReferenceSeg = Seg(TableReferenceSegment)
+    for parent_type, _, this_seg, subquery in nested_subqueries:
+        alias_name = ctes.get_name(this_seg.children(is_type("alias_expression")))
+        new_cte = CTESegment(
+            segments=(
+                CodeSegment(
+                    raw=alias_name,
+                    name="naked_identifier",
+                    type="identifier",
+                ),
+                WhitespaceSegment(),
+                segmentify("AS"),
+                WhitespaceSegment(),
+                subquery,
+            )
+        )
+        # (mypy is not understanding the generic)
+        ctes.insert_cte(new_cte)  # type: ignore
+        # TODO: Create non-mutative helper function.
+        # We will replace the whole tree, mutate the table expression.
+        this_seg[0].segments = (
+            TableExpressionSeg(
+                segments=(
+                    TableReferenceSeg(
+                        segments=(
+                            CodeSegment(
+                                raw=alias_name,
+                                name="naked_identifier",
+                                type="identifier",
+                            ),
+                        )
+                    ),
                 )
-                if seg:
-                    return LintResult(
-                        anchor=seg[0],
-                        description=f"{parent_type} clauses should not contain "
-                        "subqueries. Use CTEs instead",
-                    )
-        return None
+            ),
+        )
+        res = LintResult(
+            anchor=subquery,
+            description=f"{parent_type} clauses should not contain "
+            "subqueries. Use CTEs instead",
+            fixes=[],
+        )
+        lint_results.append(res)
+
+    if ctes.has_duplicates() or is_recursive:
+        # If we have duplicate CTE names just don't fix anything
+        return lint_results
+
+    new_select = WithCompoundStatementSegment(
+        segments=tuple(
+            [
+                segmentify("WITH"),
+                WhitespaceSegment(),
+                *ctes.get_segements(),
+                NewlineSegment(),
+                output_select[0],
+            ]
+        )
+    )
+    # Add fixes to the last result only
+    res = lint_results.pop()
+    res.fixes = [LintFix.replace(root_select[0], edit_segments=[new_select])]
+    lint_results.append(res)
+    return lint_results
+
+
+def _is_child(maybe_parent: Segments, maybe_child: Segments) -> bool:
```

Comment:
This function looks similar to `BaseSegment.path_to()`. Could you use that instead?

https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/core/parser/segments/base.py#L866L892

Reply:
I think they are quite different that other function does a recursive crawl. This just quickly checks the position markers